### DISCUSSION TITLE
How to run multiple commands in one Github Actions Docker

### DISCUSSION THREAD

**Discussion by baruchiro** [community/community#26111] [Actions]
<p><a href="https://stackoverflow.com/questions/56726429/how-to-run-multiple-commands-in-one-github-actions-docker" rel="nofollow noopener">Asked in StackOverflow</a>.</p>
<p>What is the right way for running multiple commands in one action?</p>
<p><strong>For example:</strong></p>
<p>I want to run a python script as action.</p>
<p>Before running this script I need to install the requirements.txt.</p>
<p>I can think of several options:</p>
<ul>
<li>Create a Dockerfile with the command RUN pip install -r requirements.txt in it.</li>
<li>Use the python:3 image, and run the pip install -r requirements.txt in the entrypoint.sh file before running the arguments from args in main.workflow.</li>
<li>use both pip install and python myscript.py as args</li>
</ul>
<p>Another example:</p>
<p>I want to run a script that exists in my repository, then compare 2 files (its output and a file that already exists).</p>
<p>This is a process that includes <strong>two commands</strong> , whereas in the first example, the pip install command can be considered <strong>a building command</strong> rather than a test command.</p>
<p><strong>the question:</strong></p>
<p>Can I create another Docker for another command, which will contain the output of the previous Docker?</p>
<p>I’m looking for guidelines for the location of the command in Dockerfile, in entrypoint or in args.</p>

**Comment by lee-dohm** [comment, marked as answer]
<p>There’s not necessarily a “right” or a “wrong” way to make these decisions. My personal opinion on the subjects:</p>
<ul>
<li>
<code>args</code> should be used to customize the behavior of an Action from one run to another
<ul>
<li>So if you have an Action that does something but you need it done two different ways in two separate Workflows, then an argument is probably the right choice</li>
</ul>
</li>
<li>
<code>env</code> should be used to create named values (that don’t need to be secure) that might be shared between Actions
<ul>
<li>Like if multiple Actions need to know where the <code>DOCS_DIR</code> is, this might be a valid use for an environment variable</li>
</ul>
</li>
<li>If the code in an Action requires some tool or even fixed dependencies to do its job, I would typically build that into the Docker image as a <code>RUN</code> step so that the code in <code>entrypoint.sh</code> is, as much as possible, performing the Action’s designed task</li>
<li>Much like <a href="https://en.wikipedia.org/wiki/Unix_philosophy#Origin" rel="nofollow noopener">the Unix philosophy</a> for command-line tools, each Action should do one thing well and Workflows should be used to chain multiple Actions together</li>
</ul>
<p>But these are only my personal thoughts on it from my meager experience so far and, much like designing command-line tools, there are a bunch of different ways to achieve the same ends.</p>
<p>You can see how I’ve designed one Action for a recent project of mine in <a href="https://github.com/lee-dohm/generate-elixir-docs" rel="nofollow noopener">https://github.com/lee-dohm/generate-elixir-docs</a>. I’ll probably move <a href="https://github.com/lee-dohm/generate-elixir-docs/blob/18e0376fc0b192216378fb97b0dcc4f96313f7d6/entrypoint.sh#L20-L22" rel="nofollow noopener">these lines</a> into the Dockerfile eventually, because they can be installed in the image and cached. But <a href="https://github.com/lee-dohm/generate-elixir-docs/blob/18e0376fc0b192216378fb97b0dcc4f96313f7d6/entrypoint.sh#L24-L25" rel="nofollow noopener">this line</a> needs to remain in the <code>entrypoint.sh</code> because the definition of the dependencies is stored in the repository itself and may change from one invocation of the Action to the next.</p>
<p>Again, there’s no hard and fast right or wrong answer to any of these questions. It mostly depends on how you’re intending the Action to be used: only for the one specific project or more generally across multiple projects.</p>
<p>I hope that helps!</p>

**Comment by NullVoxPopuli** [comment]
<p>I have a publish-to-dockerhub script that is in bash, and I’d prefer to use that for publishing. I don’t want to use 2+ different actions. That ties me into the “GitHub Actions” system, and I’d prefer to just set up my environment requirements, and use my script.</p>
<p>Is there a way to do that? I literally just need docker and bash installed in this GH Actions job :slight_smile:</p>
